Make sure that Divolte has a graceful shutdown

Fokko opened the pull request:
When we want to run Divolte on top of k8s, we have to make sure that it implements a graceful shutdown. In case of rolling updates, and autoscaling, it has to send all the data to the sinks before shutting down.

friso commented on src/main/java/io/divolte/server/PingHandler.java:
```diff
@@ -16,32 +16,49 @@
-final class PingHandler {
+final class PingHandler implements HttpHandler {
     private static final Logger logger = LoggerFactory.getLogger(Server.class);
+    private boolean shutdown;
```
Assuming this is not necessarily set from the thread that reads it, I think you'll want to mark that `volatile`.

Fokko replied:
@friso Nice catch

friso commented on src/main/java/io/divolte/server/PingHandler.java:
```diff
@@ -16,32 +16,49 @@
+
+            logger.info("Return 503 on health check to indicate shutdown");
+        } else {
+            exchange.getResponseSender().send("Pong", StandardCharsets.UTF_8);
```
The current ping handler sends lowercase `pong` in the response body. While unlikely, there could be deployments with health checks that check for that exact response. (Also, I have a personal preference for lowercase pongs.)

Fokko replied:
Good point Friso. I should be lowercase.

Fokko commented:
So I've addressed the comments of Friso, and started on building a test. I have separate branch for this: https://github.com/Fokko/divolte-collector/tree/fd-handle-sigterm-big

```
#!/usr/bin/env bash

export MINIKUBE_WANTUPDATENOTIFICATION=false
export MINIKUBE_WANTREPORTERRORPROMPT=false
export MINIKUBE_HOME=$HOME
export CHANGE_MINIKUBE_NONE_USER=true
export KUBECONFIG=$HOME/.kube/config

minikube addons enable ingress

# Build the Divolte container twice, with the same code to simulate
# a rolling update

docker build --no-cache -t fokkodriesprong/divolte-test:v1 .
docker push fokkodriesprong/divolte-test:v1

docker build --no-cache -t fokkodriesprong/divolte-test:v2 .
docker push fokkodriesprong/divolte-test:v2

kubectl apply -f k8s/volume.yaml

kubectl apply -f k8s/divolte.yaml

export URL="http://minikube-divolte/csc-event?p=0%3Ajejuh70g%3Ac~K96e0rIFYLOwbqteqf618YORLVF_5N&s=0%3Ajejuh70g%3AHJPj5EBtPm3K6sQJevssPlIfRzZuWxHT&v=0%3Al_5A1Sks0kBcM7APxzD1ouTcNRUjnqiz&e=0%3Al_5A1Sks0kBcM7APxzD1ouTcNRUjnqiz0&c=jejuj2p3&n=f&f=f&l=http%3A%2F%2Fminikube-divolte%2F&i=140&j=nl&k=2&w=10l&h=46&t=pageView&x=41o68f"

while ! nc -z minikube-divolte 80; do
  echo "Divolte not yet available"
  sleep 0.1 # wait for 1/10 of the second before check again
done

set -B # enable brace expansion
for i in {1..10000}; do
    STATUSCODE=$(curl --silent --output /dev/null --write-out "%{http_code}" $URL)
    if test $STATUSCODE -ne 200; then
        echo "BOOM"
    fi

    n=$((i%100))
    if test $n -eq 0; then
        echo "Step $i"
    fi
done
```

This builds two images of the same source, and then it will send a lot of requests to the two Divolte instances. In a seperate window I perform a rolling update by applying `kubectl apply -f k8s/divolte-v2.yaml`. With the master branch, I get non 200 requests:
```
Step 100
Step 200
Step 300
Step 400
Step 500
Step 600
BOOM
BOOM
BOOM
BOOM
BOOM
BOOM
BOOM
BOOM
BOOM
BOOM
BOOM
Step 700
Step 800
Step 900
Step 1000
```

On this `fd-handle-sigterm` branch, the non-200 HTTP requests are gone. So the load-balancer safely removes the Divolte instance from pool. The next step that I'll perform is checking if all the requests are still in there. I'll send 10k requests, and I expect 10k requests in the Avro files.

Fokko commented:
Running two Divolte instances:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ kubectl get pods
NAME                       READY     STATUS    RESTARTS   AGE
divolte-7db85bfc9f-7xstq   0/1       Running   0          49s
divolte-7db85bfc9f-sm66s   0/1       Running   0          49s
```
So I've created a persistent volume on the minikube vm:
```
$ find divolte/
divolte/
divolte/published
divolte/inflight
```
Created a Divolte deployment with 2 instances and mounted the volume:
```
$ find divolte/
divolte/
divolte/published
divolte/inflight
divolte/inflight/20180309151952-divolte-tracking-divolte-7db85bfc9f-7xstq-1.avro.partial
divolte/inflight/.20180309151954-divolte-tracking-divolte-7db85bfc9f-sm66s-1.avro.partial.crc
divolte/inflight/20180309151954-divolte-tracking-divolte-7db85bfc9f-sm66s-1.avro.partial
divolte/inflight/.20180309151952-divolte-tracking-divolte-7db85bfc9f-7xstq-1.avro.partial.crc
```
Then do a rolling update `kubectl apply -f k8s/divolte-v2.yaml`:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ kubectl apply -f k8s/divolte-v2.yaml
deployment "divolte" configured
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ kubectl get pods
NAME                       READY     STATUS              RESTARTS   AGE
divolte-7db85bfc9f-7xstq   1/1       Running             0          2m
divolte-7db85bfc9f-sm66s   0/1       Terminating         0          2m
divolte-86558d5776-bkgqt   0/1       ContainerCreating   0          5s
divolte-86558d5776-vrqn9   0/1       ContainerCreating   0          5s
```
We notice that one of the contains is getting shut down:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ kubectl logs -f divolte-7db85bfc9f-sm66s
2018-03-09 15:19:52.613Z [main] INFO  [Version]: HV000001: Hibernate Validator null
2018-03-09 15:19:53.483Z [main] INFO  [SchemaRegistry]: Using builtin default Avro schema.
2018-03-09 15:19:53.789Z [main] INFO  [SchemaRegistry]: Loaded schemas used for mappings: [default]
2018-03-09 15:19:53.792Z [main] INFO  [SchemaRegistry]: Inferred schemas used for sinks: [hdfs]
2018-03-09 15:19:54.146Z [main] WARN  [NativeCodeLoader]: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2018-03-09 15:19:54.550Z [main] INFO  [Server]: Initialized sinks: [hdfs]
2018-03-09 15:19:54.612Z [main] INFO  [Mapping]: Using built in default schema mapping.
2018-03-09 15:19:54.678Z [main] INFO  [UserAgentParserAndCache]: Using non-updating (resource module based) user agent parser.
2018-03-09 15:19:55.844Z [main] INFO  [UserAgentParserAndCache]: User agent parser data version: 20141024-01
2018-03-09 15:20:04.562Z [main] INFO  [Slf4jErrorManager]: 0 error(s), 0 warning(s), 87.33860891295294% typed
2018-03-09 15:20:04.565Z [main] INFO  [JavaScriptResource]: Pre-compiled JavaScript source: divolte.js
2018-03-09 15:20:04.660Z [main] INFO  [GzippableHttpBody]: Compressed resource: 9558 -> 4162
2018-03-09 15:20:04.691Z [main] INFO  [BrowserSource]: Registered source[browser] script location: /divolte.js
2018-03-09 15:20:04.691Z [main] INFO  [BrowserSource]: Registered source[browser] event handler: /csc-event
2018-03-09 15:20:04.691Z [main] INFO  [Server]: Initialized sources: [browser]
2018-03-09 15:20:04.738Z [main] INFO  [Server]: Starting server on 0.0.0.0:8290
2018-03-09 15:20:04.764Z [main] INFO  [xnio]: XNIO version 3.3.8.Final
2018-03-09 15:20:04.797Z [main] INFO  [nio]: XNIO NIO Implementation Version 3.3.8.Final
2018-03-09 15:21:36.684Z [Thread-8] WARN  [Server]: Requested to kill process, init graceful shutdown
2018-03-09 15:21:36.685Z [Thread-8] INFO  [Server]: Wait 2200ms to let upstream know that we are shutting down
2018-03-09 15:21:37.119Z [XNIO-1 I/O-1] INFO  [Server]: Return 503 on health check to indicate shutdown
2018-03-09 15:21:38.118Z [XNIO-1 I/O-1] INFO  [Server]: Return 503 on health check to indicate shutdown
2018-03-09 15:21:38.885Z [Thread-8] INFO  [Server]: Wait 192500ms till all the connections are closed
2018-03-09 15:21:38.885Z [Thread-8] INFO  [Server]: Undertow shut down with no remaining connections
2018-03-09 15:21:38.886Z [Thread-8] INFO  [Server]: Stopping thread pools.
2018-03-09 15:21:39.764Z [Thread-8] INFO  [Server]: Closing the sinks filesystem connection.
2018-03-09 15:21:40.774Z [Thread-8] INFO  [Server]: Closing HDFS filesystem connection.
rpc error: code = Unknown desc = Error: No such container: d1f33e3cdb952a96d0a044807fc8b045f2a965e52d4c36299918a95724c7c4d5
```
So the files go to published:
```
$ find divolte/
divolte/
divolte/published
divolte/published/20180309151954-divolte-tracking-divolte-7db85bfc9f-sm66s-1.avro
divolte/published/.20180309151954-divolte-tracking-divolte-7db85bfc9f-sm66s-1.avro.crc
divolte/inflight
divolte/inflight/20180309151952-divolte-tracking-divolte-7db85bfc9f-7xstq-1.avro.partial
divolte/inflight/.20180309151952-divolte-tracking-divolte-7db85bfc9f-7xstq-1.avro.partial.crc
```
And eventually the new pods are available:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ kubectl get pods
NAME                       READY     STATUS        RESTARTS   AGE
divolte-7db85bfc9f-7xstq   0/1       Terminating   0          3m
divolte-86558d5776-bkgqt   1/1       Running       0          1m
divolte-86558d5776-vrqn9   0/1       Running       0          1m
```
After finishing the load test, I delete the deployment to shut Divolte down:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ kubectl delete deployment divolte
deployment "divolte" deleted
```
And I end up with the files on the persistent disk of Kubernetes:
```
$ find divolte/
divolte/
divolte/published
divolte/published/20180309152237-divolte-tracking-divolte-86558d5776-vrqn9-1.avro
divolte/published/20180309152221-divolte-tracking-divolte-86558d5776-bkgqt-1.avro
divolte/published/20180309151954-divolte-tracking-divolte-7db85bfc9f-sm66s-1.avro
divolte/published/20180309151952-divolte-tracking-divolte-7db85bfc9f-7xstq-1.avro
divolte/published/.20180309151952-divolte-tracking-divolte-7db85bfc9f-7xstq-1.avro.crc
divolte/published/.20180309152237-divolte-tracking-divolte-86558d5776-vrqn9-1.avro.crc
divolte/published/.20180309151954-divolte-tracking-divolte-7db85bfc9f-sm66s-1.avro.crc
divolte/published/.20180309152221-divolte-tracking-divolte-86558d5776-bkgqt-1.avro.crc
```

So, when we run this without any rolling update we get exactly 10k rows:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ avro-tools tojson /tmp/divolte/divolte/published/20180309153621-divolte-tracking-divolte-7db85bfc9f-8jcbb-1.avro | wc -l
    5000
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ avro-tools tojson /tmp/divolte/divolte/published/20180309153620-divolte-tracking-divolte-7db85bfc9f-ws4d5-1.avro | wc -l
    5000
```
With the rolling update:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ avro-tools tojson /tmp/divolte/data/divolte/published/20180309154529-divolte-tracking-divolte-7db85bfc9f-cjvqc-1.avro | wc -l
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
    2094
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ avro-tools tojson /tmp/divolte/data/divolte/published/20180309154529-divolte-tracking-divolte-7db85bfc9f-w5xxc-1.avro | wc -l
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
     555
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ avro-tools tojson /tmp/divolte/data/divolte/published/20180309154746-divolte-tracking-divolte-86558d5776-598jm-1.avro | wc -l
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
    3801
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ avro-tools tojson /tmp/divolte/data/divolte/published/20180309154755-divolte-tracking-divolte-86558d5776-ww74k-1.avro | wc -l
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
    3550

In total: 2094+555+3801+3550=10k (==no loss)
```

asnare commented:
Do you happen to have the results from current `master`?

Fokko commented:
Running on master:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ kubectl logs -f divolte-7db85bfc9f-ppj4l
2018-03-09 22:56:01.686Z [main] INFO  [Version]: HV000001: Hibernate Validator null
2018-03-09 22:56:02.629Z [main] INFO  [SchemaRegistry]: Using builtin default Avro schema.
2018-03-09 22:56:02.914Z [main] INFO  [SchemaRegistry]: Loaded schemas used for mappings: [default]
2018-03-09 22:56:02.917Z [main] INFO  [SchemaRegistry]: Inferred schemas used for sinks: [hdfs]
2018-03-09 22:56:03.250Z [main] WARN  [NativeCodeLoader]: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2018-03-09 22:56:03.728Z [main] INFO  [Server]: Initialized sinks: [hdfs]
2018-03-09 22:56:03.809Z [main] INFO  [Mapping]: Using built in default schema mapping.
2018-03-09 22:56:03.834Z [main] INFO  [UserAgentParserAndCache]: Using non-updating (resource module based) user agent parser.
2018-03-09 22:56:04.775Z [main] INFO  [UserAgentParserAndCache]: User agent parser data version: 20141024-01
2018-03-09 22:56:13.123Z [main] INFO  [Slf4jErrorManager]: 0 error(s), 0 warning(s), 87.33860891295294% typed
2018-03-09 22:56:13.125Z [main] INFO  [JavaScriptResource]: Pre-compiled JavaScript source: divolte.js
2018-03-09 22:56:13.172Z [main] INFO  [GzippableHttpBody]: Compressed resource: 9558 -> 4162
2018-03-09 22:56:13.188Z [main] INFO  [BrowserSource]: Registered source[browser] script location: /divolte.js
2018-03-09 22:56:13.189Z [main] INFO  [BrowserSource]: Registered source[browser] event handler: /csc-event
2018-03-09 22:56:13.189Z [main] INFO  [Server]: Initialized sources: [browser]
2018-03-09 22:56:13.209Z [main] INFO  [Server]: Starting server on 0.0.0.0:8290
2018-03-09 22:56:13.219Z [main] INFO  [xnio]: XNIO version 3.3.8.Final
2018-03-09 22:56:13.237Z [main] INFO  [nio]: XNIO NIO Implementation Version 3.3.8.Final
2018-03-09 22:57:07.198Z [Thread-8] INFO  [Server]: Stopping HTTP server.
2018-03-09 22:57:07.200Z [Thread-8] INFO  [Server]: Stopping thread pools.
2018-03-09 22:57:08.626Z [Thread-8] INFO  [Server]: Closing HDFS filesystem connection.
rpc error: code = Unknown desc = Error: No such container: 550931c346f42c54c510efb11f38208eadfd1fd08f358fb32f38d8d02802377d
```
With the patch I get only 200's, but when running `master`, I also get 502:
```
Fokkos-MacBook-Pro:divolte-collector fokkodriesprong$ for i in {1..10000}; do     STATUSCODE=$(curl --silent --output /dev/null --write-out "%{http_code}" $URL);     if test $STATUSCODE -ne 200; then         echo "BOOM $STATUSCODE";     fi;      n=$((i%100));     if test $n -eq 0; then         echo "Step $i";     fi; done
Step 100
Step 200
Step 300
Step 400
Step 500
Step 600
Step 700
Step 800
Step 900
Step 1000
Step 1100
Step 1200
Step 1300
Step 1400
Step 1500
Step 1600
Step 1700
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
BOOM 502
Step 1800
Step 1900
Step 2000
Step 2100
Step 2200
...
Step 10000
```
After SCP'ing from the minikube VM:
```
Fokkos-MacBook-Pro:published fokkodriesprong$ ls -lah .
total 5240
drwxr-xr-x  10 fokkodriesprong  wheel   320B Mar 10 00:01 .
drwxrwxrwx   4 fokkodriesprong  wheel   128B Mar 10 00:01 ..
-rw-r--r--   1 fokkodriesprong  wheel   3.6K Mar 10 00:01 .20180309225601-divolte-tracking-divolte-7db85bfc9f-wvvfn-1.avro.crc
-rw-r--r--   1 fokkodriesprong  wheel   1.1K Mar 10 00:01 .20180309225603-divolte-tracking-divolte-7db85bfc9f-ppj4l-1.avro.crc
-rw-r--r--   1 fokkodriesprong  wheel   7.8K Mar 10 00:01 .20180309225715-divolte-tracking-divolte-86558d5776-kpt5q-1.avro.crc
-rw-r--r--   1 fokkodriesprong  wheel   7.8K Mar 10 00:01 .20180309225717-divolte-tracking-divolte-86558d5776-5fzsn-1.avro.crc
-rw-r--r--   1 fokkodriesprong  wheel   457K Mar 10 00:01 20180309225601-divolte-tracking-divolte-7db85bfc9f-wvvfn-1.avro
-rw-r--r--   1 fokkodriesprong  wheel   142K Mar 10 00:01 20180309225603-divolte-tracking-divolte-7db85bfc9f-ppj4l-1.avro
-rw-r--r--   1 fokkodriesprong  wheel   997K Mar 10 00:01 20180309225715-divolte-tracking-divolte-86558d5776-kpt5q-1.avro
-rw-r--r--   1 fokkodriesprong  wheel   992K Mar 10 00:01 20180309225717-divolte-tracking-divolte-86558d5776-5fzsn-1.avro
Fokkos-MacBook-Pro:published fokkodriesprong$ avro-tools tojson 20180309225601-divolte-tracking-divolte-7db85bfc9f-wvvfn-1.avro | wc -l
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
    1752
Fokkos-MacBook-Pro:published fokkodriesprong$ avro-tools tojson 20180309225603-divolte-tracking-divolte-7db85bfc9f-ppj4l-1.avro | wc -l
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
     540
Fokkos-MacBook-Pro:published fokkodriesprong$ avro-tools tojson 20180309225715-divolte-tracking-divolte-86558d5776-kpt5q-1.avro | wc -l
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
    3830
Fokkos-MacBook-Pro:published fokkodriesprong$ avro-tools tojson 20180309225717-divolte-tracking-divolte-86558d5776-5fzsn-1.avro | wc -l
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.lib.MutableMetricsFactory).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
    3811

1752 + 540 + 3830 + 3811 = 9933 (77 lost) :'(
```

Fokko commented:
@asnare Do I need to resolve comflicts?
